File: hw4/practice.cpp
// HASK17
// a)
#include <vector>
#include <iostream>
#include <queue>
using namespace std;

int longestRun(const vector<bool>& lst){
    int max_count = 0;
    int count = 0;
    for (int i = 0; i < lst.size(); i++){
        if (lst[i]){
            count++;
            if (count > max_count){
                max_count = count;
            }
        }
        else{
            count = 0;
        }
    }
    return max_count;
}
class Tree {
public:
  unsigned value;
  vector<Tree *> children;

  Tree(unsigned value, vector<Tree *> children) {
    this->value = value;
    this->children = children;
  }
};

// HASK17
// c)
unsigned maxTreeValue(Tree* root){
    if (root == nullptr){
        return 0;
    }
    unsigned max = root->value;
    queue<Tree*> bfs;
    vector<Tree*> children = root->children;
    for (int i = 0; i < children.size(); i++){
        bfs.push(children[i]);
    }
    //iterate through bfs
    while (!bfs.empty()){
        Tree* ptr = bfs.front();
        children = ptr->children;
        bfs.pop();
        if (ptr != nullptr){
            if (ptr->value > max){
                max = ptr->value;
            }
            //add children to queue
            for (int i = 0; i < children.size(); i++){
                bfs.push(children[i]);
            }
        }
    }
    return max;
}

int main(){
    /*
    vector<bool> vec1 = {true, true, false, true, true, true, false};
    vector<bool> vec2 = {true, false, true, true};
    cout << longestRun(vec1) << endl;
    cout << longestRun(vec2) << endl;
    */
    Tree *root = new Tree(5, {
        new Tree(3, {
            new Tree(1, {}),
            new Tree(4, {})
        }),
        new Tree(7, {
            new Tree(6, {}),
            new Tree(9, {})
        })
    });
    unsigned maxValue = maxTreeValue(root);
    cout << "The maximum value in the tree is: " << maxValue << endl;
}

File: hw4/practice.hs
-- HASK17
-- b)
longest_run :: [Bool] -> Integer
longest_run lst = helper lst 0 0
    where
        helper [] total curr = max total curr
        helper (x:xs) total curr
            | x = helper xs total (curr + 1)
            | otherwise = helper xs (max total curr) 0

-- HASK17
-- d)
max_tree_value :: Tree -> Integer
max_tree_value Empty = 0
max_tree_value (Node val []) = val
max_tree_value (Node val children) = max val (maximum (map max_tree_value children))

data Tree = 
    Empty | 
    Node Integer [Tree]

-- HASK18
data Event = Travel Integer | Fight Integer | Heal Integer
super_giuseppe :: [Event] -> Integer
super_giuseppe lst = helper lst 100 False
    where
        helper [] health _ = health -- done with game
        helper _ health _ 
            | health <= 0 = -1 -- lose all hp
        helper (Travel dist:xs) health inDefenseMode
            | inDefenseMode = helper xs health inDefenseMode -- don't gain any health
            | otherwise = helper xs (min 100 (health + dist `div` 4)) inDefenseMode
        helper (Fight hplost:xs) health inDefenseMode
            | inDefenseMode = helper xs (health - hplost `div` 2) inDefenseMode -- lose half
            | otherwise = helper xs (health - hplost) (health - hplost <= 40) 
        helper (Heal hpgain:xs) health inDefenseMode = helper xs (min 100 (health + hpgain)) (health + hpgain <= 40) -- gain health and check if inDefenseMode

-- HASK19
-- a)
sumSquares :: Integer -> Integer
sumSquares 0 = 0
sumSquares n = n^2 + sumSquares(n-1)

-- b)
sumSquaresTail :: Integer -> Integer
sumSquaresTail n = helper n 0
    where
        helper 0 acc = acc
        helper n acc = helper (n-1) (acc + n^2)